Paraswap Implementation Doc (#528)
UMAprotocol/UMIPs: Implementations/paraswap-volume.md

## Title

Paraswap Volume KPI Option Calculation

## Summary

ParaSwap aggregates decentralized exchanges and other DeFi services in one comprehensive interface to streamline and facilitate users' interactions with decentralized finance on Ethereum and EVM-compatible chains. The Paraswap network wishes to incentivize increased volume to the network by offering the Paraswap community a linear option that targets an increase in trading volume.

## Intended Ancillary Data

```

> Metric:Paraswap trade volume,

Reviewer: measured in USD?

Method:"https://github.com/UMAprotocol/UMIPs/blob/master/Implementations/paraswap-volume.md",

> Aggregation:Increase in cumulative volume for the provided time range,

Reviewer: This would only be relevant if API returned cumulative volume since inception at requested timestamp, but below methodology already filters the trades within required time range.

StartTimestamp:<start_timestamp>,
StartTWAP:<start_timestamp>,
EndTWAP:<end_timestamp>,
Rounding:0
```

## Implementation

Subgraphs:

- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-fantom
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-avalanche
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-bsc
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-polygon

Paraswap wishes to track the total increase in cumulative volume over a period of time. The query that should be used is:

```
{
  swaps(first: 1000, orderBy: timestamp, orderDirection: desc, where:{timestamp_lte: timestamp.number}) {
    srcToken
    destToken
    srcAmount
    destAmount
    timestamp
  }
}
```

1. For each of the listed `Subgraphs` endpoints above, replace the `timestamp.number` value with the request timestamp and run the swaps query shown above. Note: If necessary, use the skip variable `skip:1000` to iterate through a loop to query 1,000 swaps each time. Continually skip by 1,000 until all swap responses are returned.

> 2. From the returned results of step 1, filter out returned values where values have a `timestamp` value that is less than the `StartTimestamp` value. Record each returned `srcAmount`, `srcToken`, `destAmount`, `destToken` values returned.

Reviewer: Would this be possible to achieve by doing filtering in the query? E.g. adding `timestamp_gte: StartTimestamp` to the `where` object.

Reviewer: As query has been updated you can remove filtering out from Step 2.

> 3. Group all `srcToken` and `destToken` values returned from step 2 and sum the values of `srcAmount` and `destAmount`.

Reviewer: ```suggestion
3. Group all `srcToken` and `destToken` values returned from step 2 and sum the values of `srcAmount` and `destAmount` separately.
```

Just to avoid possible confusion that one might add `destAmount` to `srcAmount`

> 4. Scale down raw amounts calculated in Step 3 by each ERC-20 token contract decimals (or convert from wei for Ether).

Reviewer: Add instructions on how to determine the native token. I would assume its `0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee`, but need to double check if this applies to all networks (e.g. is it MATIC on Polygon?)

> 5. For all swapped ERC-20 tokens / Ether get the historical pricing series from CoinGecko for the evaluation period:

Reviewer: ```suggestion
5. For all swapped ERC-20 tokens / native token get the historical pricing series from CoinGecko for the evaluation period:
```

   - Example API request for WETH token on Polygon priced in USD: https://api.coingecko.com/api/v3/coins/polygon-pos/contract/0x7ceb23fd6bc0add59e62ac25578270cff1b9f619/market_chart/range?vs_currency=usd&from=1646092800&to=1648771200
   - Locate the `prices` key value from CoinGecko API response - it should contain a list of [ timestamp, price ] value pairs. Note that returned CoinGecko timestamps are in milliseconds;
   - In case returned pricing interval is more frequent than 1 hour voters should extend the requested time range. This is necessary to have consistent pricing results independent on when the voters are calculating them.
   - Voters should verify that obtained price results agree with broad market consensus.

Sum the returned token prices and divide by the number of price values. This value will be used for the average token price for swaps.

6. Multiply the values returned from steps 4 and 5.

> 5. Sum all returned values from step 6. Round the value to 0 decimal places (e.g. 25.123 -> 25).

Reviewer: Shouldn't this be divided by 2 since calculation above considers the value in USD both for `srcAmount` and `destAmount`?

Reviewer: Also in case there is difficulty on obtaining Coingecko pricing for all tokens the fallback could use USD value of only `srcAmount` or `destAmount` whichever is available. Though this would require grouping trades by all combinations of `srcToken` and `destToken` pairs.

An example implementation script is provided [here](https://github.com/abg4/paraswap), but voters are encouraged to independently verify its calculation logic and results.

## Intended Application

> It is intended to deploy the documented KPI options on Ethereum network using [LSP contract](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/long-short-pair/LongShortPair.sol) with `General_KPI` price identifier approved in [UMIP-117](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-117.md). The contract would use [Linear LSP FPL](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/LinearLongShortPairFinancialProductLibrary.sol) with the `lowerBound` set to 0 and `upperBound` set to 1.

Reviewer: Shouldn't the `upperBound` be set to `1,000,000,000` scaled to 18 decimals to represent $1 billion?

## Intended Ancillary Data

```
Metric:Paraswap trade volume measured in USD,
Method:"https://github.com/UMAprotocol/UMIPs/blob/master/Implementations/paraswap-volume.md",
StartTimestamp:<start_timestamp>,
StartTWAP:<start_timestamp>,
EndTWAP:<end_timestamp>,
Rounding:0
```

## Implementation

Subgraphs:

- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-fantom
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-avalanche
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-bsc
- https://api.thegraph.com/subgraphs/name/paraswap/paraswap-subgraph-polygon

Paraswap wishes to track the total increase in cumulative volume over a period of time. The query that should be used is:

```
{
  swaps(first: 1000, orderBy: timestamp, orderDirection: desc, where:{timestamp_lte: timestamp.number, timestamp_gte: <StartTimestamp>}) {
    srcToken
    destToken
    srcAmount
    destAmount
    timestamp
  }
}
```

> 1. For each of the listed `Subgraphs` endpoints above, replace the `timestamp.number` value with the request timestamp and run the swaps query shown above. Note: If necessary, use the skip variable `skip:1000` to iterate through a loop to query 1,000 swaps each time. Continually skip by 1,000 until all swap responses are returned.

Reviewer: ```suggestion
1. For each of the listed `Subgraphs` endpoints above, replace the `timestamp.number` value with the request timestamp, use `StartTimestamp` from ancillary data and run the swaps query shown above. Note: If necessary, use the skip variable `skip:1000` to iterate through a loop to query 1,000 swaps each time. Continually skip by 1,000 until all swap responses are returned.
```
